MockitoJUnitRunner.StrictStubs does not detect 'Unnecessary Stubbing' when inline mock maker is enabled

- mockito-core 2.7.13
- OpenJDK 1.8.0_121
- Ubuntu 16.10
- Enable `mock-maker-inline` (or use `mockito-inline` artifact)

If running the following test, 'Unnecessary Stubbing' is not detected.

```java
@RunWith(MockitoJUnitRunner.StrictStubs.class)
public class MockitoTest {

    @Mock
    ArrayList<String> mock;

    @Test
    public void should_pass() {
        when(mock.get(0)).thenReturn("foo");
        assertEquals("foo", mock.get(0));
    }

    @Test
    public void should_detect_unnecessary_stubbing() {
        when(mock.get(1)).thenReturn("foo");
    }

}
```

In the above code, 'Unnecessary Stubbing' is detected by changing ArrayList to List.
```java

    @Mock
//    ArrayList<String> mock;
    List<String> mock;

```

This problem does not occur with default mock maker.

I wrote a new test for this issue.

```java
public class StrictStubsRunnerByMockingAbstractClassTest {

    private final JUnitCore jUnitCore = new JUnitCore();

    @Test
    public void test_for_abstract_method_should_detect_unnecessary_stubbings() {
        runTest(ForAbstractMethod.class);
    }

    @Test
    public void test_for_concrete_method_should_detect_unnecessary_stubbings() {
        runTest(ForConcreteMethod.class);
    }

    private void runTest(Class<?> testClass) {
        Result result = jUnitCore.run(testClass);
        List<Failure> failures = result.getFailures();
        assertThat(failures.size(), equalTo(1));
        Failure failure = failures.get(0);
        assertThat(failure.getException(), instanceOf(UnnecessaryStubbingException.class));
    }

    static abstract class TestTarget {

        abstract String abstractMethod();

        String concreteMethod() {
            throw new UnsupportedOperationException();
        }

    }

    @RunWith(MockitoJUnitRunner.StrictStubs.class)
    public static class ForAbstractMethod {

        @Mock
        TestTarget mock;

        @Test
        public void correct_stubbing() {
            when(mock.abstractMethod()).thenReturn("");
            mock.abstractMethod();
        }

        @Test
        public void unused_stubbing() {
            when(mock.abstractMethod()).thenReturn("");
        }

    }

    @RunWith(MockitoJUnitRunner.StrictStubs.class)
    public static class ForConcreteMethod {

        @Mock
        TestTarget mock;

        @Test
        public void correct_stubbing() {
            when(mock.concreteMethod()).thenReturn("");
            mock.concreteMethod();
        }

        @Test
        public void unused_stubbing() {
            when(mock.concreteMethod()).thenReturn("");
        }

    }

}
```

To test this test on the master, I put this into the following directory.
1. src/test/java/org/mockitousage/junitrunner (using the default mock maker)
1. inline/src/test/java/org/mockitoinline (using the inline mock maker)

In the first case, this test passed.
```
$ ./gradlew -Dorg.gradle.parallel=false clean :test --tests "org.mockitousage.junitrunner.StrictStubsRunnerByMockingAbstractClassTest"
Version: 2.7.14
:clean
:android:clean UP-TO-DATE
:extTest:clean UP-TO-DATE
:inline:clean
:testng:clean UP-TO-DATE
:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:copyMockMethodDispatcher
:processResources UP-TO-DATE
:classes
:createTestResources
:compileTestJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:removeTestResources
:processTestResources UP-TO-DATE
:testClasses
:test

BUILD SUCCESSFUL

Total time: 5.707 secs
```

However, in the second case it failed.
```
$ ./gradlew -Dorg.gradle.parallel=false clean :inline:test --tests "org.mockitoinline.StrictStubsRunnerByMockingAbstractClassTest"
Version: 2.7.14
:clean
:android:clean UP-TO-DATE
:extTest:clean UP-TO-DATE
:inline:clean
:testng:clean UP-TO-DATE
:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:copyMockMethodDispatcher
:processResources UP-TO-DATE
:classes
:jar
:inline:compileJava UP-TO-DATE
:inline:processResources
:inline:classes
:inline:compileTestJava
:inline:processTestResources UP-TO-DATE
:inline:testClasses
:inline:test

org.mockitoinline.StrictStubsRunnerByMockingAbstractClassTest > test_for_concrete_method_should_detect_unnecessary_stubbings FAILED
    java.lang.AssertionError: 
    Expected: <1>
         but: was <0>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.mockitoinline.StrictStubsRunnerByMockingAbstractClassTest.runTest(StrictStubsRunnerByMockingAbstractClassTest.java:36)
        at org.mockitoinline.StrictStubsRunnerByMockingAbstractClassTest.test_for_concrete_method_should_detect_unnecessary_stubbings(StrictStubsRunnerByMockingAbstractClassTest.java:30)

2 tests completed, 1 failed
:inline:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':inline:test'.
> There were failing tests. See the report at: file:///home/tmura/IdeaProjects/mockito/subprojects/inline/build/reports/tests/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 3.505 secs

```

For the abstract method, 'Unnecessary Stubbing' was detected correctly, but it did not detected for the concrete method.

--- a/src/main/java/org/mockito/internal/creation/bytebuddy/InterceptedInvocation.java
+++ b/src/main/java/org/mockito/internal/creation/bytebuddy/InterceptedInvocation.java
@@ -42,14 +42,15 @@ class InterceptedInvocation implements Invocation, VerificationAwareInvocation {
                                  MockitoMethod mockitoMethod,
                                  Object[] arguments,
                                  SuperMethod superMethod,
+                                 Location location,
                                  int sequenceNumber) {
         this.mock = mock;
         this.mockitoMethod = mockitoMethod;
         this.arguments = ArgumentsProcessor.expandVarArgs(mockitoMethod.isVarArgs(), arguments);
         this.rawArguments = arguments;
         this.superMethod = superMethod;
+        this.location = location;
         this.sequenceNumber = sequenceNumber;
-        location = new LocationImpl();
     }
 
     @Override

--- a/src/main/java/org/mockito/internal/creation/bytebuddy/MockMethodAdvice.java
+++ b/src/main/java/org/mockito/internal/creation/bytebuddy/MockMethodAdvice.java
@@ -9,6 +9,7 @@ import net.bytebuddy.implementation.bind.annotation.Argument;
 import net.bytebuddy.implementation.bind.annotation.This;
 import net.bytebuddy.implementation.bytecode.assign.Assigner;
 import org.mockito.exceptions.base.MockitoException;
+import org.mockito.internal.debugging.LocationImpl;
 import org.mockito.internal.exceptions.stacktrace.ConditionalStackTraceFilter;
 import org.mockito.internal.invocation.SerializableMethod;
 import org.mockito.internal.util.concurrent.WeakConcurrentMap;
@@ -21,6 +22,8 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.concurrent.Callable;
 
 public class MockMethodAdvice extends MockMethodDispatcher {
@@ -91,10 +94,13 @@ public class MockMethodAdvice extends MockMethodDispatcher {
         } else {
             superMethod = new SuperMethodCall(selfCallInfo, origin, instance, arguments);
         }
+        Throwable t = new Throwable();
+        t.setStackTrace(skipInlineMethodElement(t.getStackTrace()));
         return new ReturnValueWrapper(interceptor.doIntercept(instance,
                 origin,
                 arguments,
-                superMethod));
+                superMethod,
+                new LocationImpl(t)));
     }
 
     @Override
@@ -198,6 +204,21 @@ public class MockMethodAdvice extends MockMethodDispatcher {
             new ConditionalStackTraceFilter().filter(hideRecursiveCall(cause, new Throwable().getStackTrace().length, origin.getDeclaringClass()));
             throw cause;
         }
+    }
+
+    // With inline mocking, mocks for concrete classes are not subclassed, so elements of the stubbing methods are not filtered out.
+    // Therefore, if the method is inlined, skip the element.
+    private static StackTraceElement[] skipInlineMethodElement(StackTraceElement[] elements) {
+        List<StackTraceElement> list = new ArrayList<StackTraceElement>(elements.length);
+        for (int i = 0; i < elements.length; i++) {
+            StackTraceElement element = elements[i];
+            list.add(element);
+            if (element.getClassName().equals(MockMethodAdvice.class.getName()) && element.getMethodName().equals("handle")) {
+                // If the current element is MockMethodAdvice#handle(), the next is assumed to be an inlined method.
+                i++;
+            }
+        }
+        return list.toArray(new StackTraceElement[list.size()]);
     }
 
     private static class ReturnValueWrapper implements Callable<Object> {

--- a/src/main/java/org/mockito/internal/creation/bytebuddy/MockMethodInterceptor.java
+++ b/src/main/java/org/mockito/internal/creation/bytebuddy/MockMethodInterceptor.java
@@ -7,9 +7,11 @@ package org.mockito.internal.creation.bytebuddy;
 import net.bytebuddy.implementation.bind.annotation.*;
 import org.mockito.internal.InternalMockHandler;
 import org.mockito.internal.creation.DelegatingMethod;
+import org.mockito.internal.debugging.LocationImpl;
 import org.mockito.internal.invocation.MockitoMethod;
 import org.mockito.internal.invocation.SerializableMethod;
 import org.mockito.internal.progress.SequenceNumber;
+import org.mockito.invocation.Location;
 import org.mockito.invocation.MockHandler;
 import org.mockito.mock.MockCreationSettings;
 
@@ -38,12 +40,27 @@ public class MockMethodInterceptor implements Serializable {
                        Method invokedMethod,
                        Object[] arguments,
                        InterceptedInvocation.SuperMethod superMethod) throws Throwable {
-        return handler.handle(new InterceptedInvocation(
+        return doIntercept(
                 mock,
-                createMockitoMethod(invokedMethod),
+                invokedMethod,
                 arguments,
                 superMethod,
-                SequenceNumber.next()
+                new LocationImpl()
+        );
+    }
+
+    Object doIntercept(Object mock,
+                       Method invokedMethod,
+                       Object[] arguments,
+                       InterceptedInvocation.SuperMethod superMethod,
+                       Location location) throws Throwable {
+        return handler.handle(new InterceptedInvocation(
+            mock,
+            createMockitoMethod(invokedMethod),
+            arguments,
+            superMethod,
+            location,
+            SequenceNumber.next()
         ));
     }
 

--- a/src/main/java/org/mockito/internal/debugging/LocationImpl.java
+++ b/src/main/java/org/mockito/internal/debugging/LocationImpl.java
@@ -19,8 +19,16 @@ public class LocationImpl implements Location, Serializable {
     }
 
     public LocationImpl(StackTraceFilter stackTraceFilter) {
+        this(stackTraceFilter, new Throwable());
+    }
+
+    public LocationImpl(Throwable stackTraceHolder) {
+        this(new StackTraceFilter(), stackTraceHolder);
+    }
+
+    private LocationImpl(StackTraceFilter stackTraceFilter, Throwable stackTraceHolder) {
         this.stackTraceFilter = stackTraceFilter;
-        stackTraceHolder = new Throwable();
+        this.stackTraceHolder = stackTraceHolder;
     }
 
     @Override

--- a/subprojects/inline/src/test/java/org/mockitoinline/StubbingLocationTest.java
+++ b/subprojects/inline/src/test/java/org/mockitoinline/StubbingLocationTest.java
@@ -1,0 +1,34 @@
+package org.mockitoinline;
+
+import java.util.Collections;
+import java.util.Set;
+import org.junit.Test;
+import org.mockito.internal.invocation.finder.AllInvocationsFinder;
+import org.mockito.stubbing.Stubbing;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class StubbingLocationTest {
+
+    @Test
+    public void stubbing_location_should_be_the_correct_point() {
+        ConcreteClass mock = mock(ConcreteClass.class);
+        String frame;
+        // Initializing 'frame' at the method parameter point is to gain the exact line number of the stubbing point.
+        when(mock.concreteMethod(frame = Thread.currentThread().getStackTrace()[1].toString())).thenReturn("");
+        mock.concreteMethod(frame);
+        Set<Stubbing> stubbings = AllInvocationsFinder.findStubbings(Collections.singleton(mock));
+        assertEquals(1, stubbings.size());
+        String location = stubbings.iterator().next().getInvocation().getLocation().toString();
+        assertEquals("-> at " + frame, location);
+    }
+
+    static final class ConcreteClass {
+        String concreteMethod(String s) {
+            throw new RuntimeException(s);
+        }
+    }
+
+}

